aws/lambda/log-classifier/src/engine.rs

```diff
@@ -44,3 +46,312 @@ pub fn evaluate_ruleset(ruleset: &RuleSet, log: &Log) -> Option<Match> {
         .flat_map(|rule| evaluate_rule(rule, log))
         .max_by(|a, b| a.rule.priority.cmp(&b.rule.priority))
 }
+
+/// Evaluates a ruleset against all lines of a log, returning matching line numbers.
+///
+/// # Arguments
+///
+/// * `ruleset` - A reference to the RuleSet to evaluate.
+/// * `log` - A reference to the Log to evaluate against.
+///
+/// # Returns
+///
+/// A vector of line numbers that match any rule in the ruleset.
+/// Note: Each rule in ruleset can only match 1 line which is the last line that matches the rule.
+pub fn evaluate_ruleset_all_lines(ruleset: &RuleSet, log: &Log) -> Vec<usize> {
+    let matches: Vec<Match> = ruleset
+        .rules
+        .par_iter()
+        .flat_map(|rule| evaluate_rule(rule, log))
+        .collect();
+    // combine captures into a single vector
+    let mut lines = Vec::new();
+    for m in matches {
+        lines.push(m.line_number);
+    }
+    lines
+}
+
+/// Merges line numbers into chunks based on specified size constraints.
+///
+/// # Arguments
+///
+/// * `log` - A reference to the Log structure containing the lines.
+/// * `line_numbers` - A vector of line numbers to be chunked.
+/// * `min_chunk_size` - The minimum size of each chunk.
+/// * `max_chunk_size` - The maximum size of each chunk.
+///
+/// # Returns
+///
+/// A vector of tuples, where each tuple contains the start and end line numbers of a chunk.
+///
+/// # Panics
+///
+/// This function will panic if:
+/// * `min_chunk_size` is greater than or equal to `max_chunk_size`.
+/// * Any line number is less than 1 or greater than the number of lines in the log.
+fn get_line_number_chunks(
+    log: &Log,
+    line_numbers: Vec<usize>,
+    min_chunk_size: usize,
+    max_chunk_size: usize,
+) -> Vec<(usize, usize)> {
+    // throw user error if min_chunk_size is greater than max_chunk_size
+    if min_chunk_size >= max_chunk_size {
+        panic!("min_chunk_size cannot be greater or equal to max_chunk_size");
+    }
+
+    // check if empty
+    if line_numbers.is_empty() {
+        return Vec::new();
+    }
+
+    // sort line numbers
+    let mut line_numbers = line_numbers;
+    line_numbers.sort();
+
+    // assert length of log is at least last line number and no negative line numbers
+    if *line_numbers.first().unwrap() < 1 || *line_numbers.last().unwrap() > log.lines.len() {
+        panic!("line numbers must be within the range of the log");
+    }
+
+    // merge line numbers into chunks
+    let mut merged_line_numbers = Vec::new();
+    let mut chunk = Vec::new();
+    // as we add min_chunk_size / 2 to both sides
+    let effective_max_chunk_size = max_chunk_size - min_chunk_size;
+    for line_number in line_numbers {
+        if chunk.is_empty() {
+            chunk.push(line_number);
+        } else if line_number - chunk.first().unwrap() < effective_max_chunk_size {
+            chunk.push(line_number);
+        } else {
+            let start = max(chunk.first().unwrap() - min_chunk_size / 2, 1);
```

Comment:
Hm, I think this logic is not sufficiently tested.

1. **Major bug**: imagine `min_chunk_size` and `max_chunk_size` being close, e.g. effective_max_chunk_size = 1 (or even 0 if you fix panic condition).  And let's say `min_chunk_size = 10`.

   If line numbers are also close, like [1, 3, 5], they will be added as separate (overlapping) chunks, instead of a single chunk.

2. When `start` is capped at `1`, `end` is not extended accordingly, meaning first chunk can be smaller than `min_chunk_size`. Same with `end` capping.

3. minor, but `min_chunk_size / 2` will shorten resulting chunk by 1 when `min_chunk_size` is odd.

Also, do we need to store all lines in the `chunk`? It seems like a pair of `chunk_start` and `chunk_end` is sufficient (that's what is added to `merged_line_numbers`).

Reply:
Here's a pseudocode that should fix the issues:

```
chunk_start = chunk_end = -1

for line in line_numbers:

    if chunk_start == -1:
        chunk_start = chunk_end = line
    elif line - chunk_end <= max_chunk_size:
        chunk_end = line
    else: # > max_chunk_size
        mid_point = (chunk_start + chunk_end) // 2
        chunk_start = max(min(chunk_start, mid_point - min_chunk_size // 2), 1)
        chunk_end = min(max(chunk_end, mid_point + min_chunk_size // 2 + min_chunk_size % 2), max_line_number)

        if chunk_end - chunk_start < min_chunk_size:
            chunk_end = min(chunk_start + min_chunk_size, max_line_number)
            chunk_start = max(chunk_end - min_chunk_size, 1)
            
        # process chunk
```